lib/RecipeExecutor.jsm

```diff
@@ -0,0 +1,1075 @@
+      filter_by_value: this.filterByValue,
+      l2_normalize: this.l2Normalize,
+      prob_normalize: this.probNormalize,
+      set_default: this.setDefault,
+      lookupValue: this.lookupValue,
+      copy_to_map: this.copyToMap,
+      scalar_multiply_tag: this.scalarMultiplyTag,
+      apply_softmax_tags: this.applySoftmaxTags
+    };
+    this.ITEM_COMBINER_REGISTRY = {
+      combiner_add: this.combinerAdd,
+      combiner_max: this.combinerMax,
+      combiner_collect_values: this.combinerCollectValues
+    };
+    this.nbTaggers = nbTaggers;
+    this.nmfTaggers = nmfTaggers;
+  }
+
+  /**
+   * Determines the type of a field. Valid types are:
+   *   string
+   *   number
+   *   array
+   *   map (strings to anything)
+   */
+  _typeOf(data) {
+    let t = typeof(data);
+    if (t === "object") {
+      if (Array.isArray(data)) {
+        return "array";
+      }
+      return "map";
+    }
+    return t;
+  }
+
+  /**
+   * Returns returns a scalar, either by because it was a constant, or by
+   * looking it up from the item. Allows for a default value if the lookup
+   * fails.
+   */
+  _lookupScalar(item, k, dfault) {
+    if (this._typeOf(k) === "number") {
+      return k;
+    } else if ((this._typeOf(k) === "string") && (k in item) && (this._typeOf(item[k]) === "number")) {
+      return item[k];
+    }
+    return dfault;
+  }
+
+  /**
+   * Simply appends all the strings from a set fields together. If the field
+   * is a list, then the cells of the list are append.
+   */
+  _assembleText(item, fields) {
+    let textArr = [];
+    for (let field of fields) {
+      if (field in item) {
+        let type = this._typeOf(item[field]);
+        if (type === "string") {
+          textArr.push(item[field]);
+        } else if (type === "array") {
+          for (let ele of item[field]) {
+            textArr.push(ele);
+          }
+        } else {
+          textArr.push(String(item[field]));
+        }
+      }
+    }
+    return textArr.join(" ");
+  }
+
+  /**
+   * Runs the naive bayes text taggers over a set of text fields. Stores the
+   * results in new fields:
+   *  nb_tags:         a map of text strings to probabilites
+   *  nb_tokens:       the tokenized text that was tagged
+   *
+   * Config:
+   *  fields:          an array containing a list of fields to concatenate and tag
+   */
+  naiveBayesTag(item, config) {
+    let text = this._assembleText(item, config.fields);
+    let tokens = tokenize(text);
+    let tags = {};
+
+    for (let nbTagger of this.nbTaggers) {
+      let result = nbTagger.tagTokens(tokens);
+      if ((result.label !== null) && result.confident) {
+        tags[result.label] = result;
+      }
+    }
+    item.nb_tags = tags;
+    item.nb_tokens = tokens;
+    return item;
+  }
+
+  /**
+   * Selectively runs NMF text taggers depending on which tags were found
+   * by the naive bayes taggers. Writes the results in into new fields:
+   *  nmf_tags_parent_weights:
```

Comment:
Let's complete the description of those arguments.

Reply:
added.